lib/luaxtable.c

```diff
@@ -0,0 +1,241 @@
+* the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_xtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+	} hook;
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static void luaxtable_release(void *private);
+static int luaxtable_reg_match(lua_State *L);
+static int luaxtable_reg_target(lua_State *L);
+
+/* Verifies the type of "opt" in the table at "idx" and returns true if the value can be extracted.
+ * If the value is not required, it returns false.
+ * If the value is required and the type is invalid, it raises an error.
+ *
+ * TODO: create a lunatik_checkfield() API to check the type of a table field and
+ * 	raise a proper error message
+ * */
+static bool luaxtable_reg_opt(lua_State *L, int idx, const char *opt, int type, bool req)
+{
+	lua_getfield(L, idx, opt);
+	int type_ = lua_type(L, -1);
+	if (type_ != type) {
+		// LUA_TNIL is 0 and LUA_TNONE is -1
+		if (type_ <= 0 && !req) {
+			/* the value was not required hence, should not try to extract it */
+			return false;
+		}
+		luaL_error(L, "invalid type for %s, expected %s", opt, lua_typename(L, type));
+	}
+
+	/* true if the value can be extracted (for optional parameters) */
+	return true;
+};
+
+#define LUAXTABLE_REG(mode, ttype)						\
+static int luaxtable_reg_##mode(lua_State *L) {				\
+	lunatik_object_t *object;					\
+	luaxtable_t *xtable_##mode;					\
+									\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	luaxtable_reg_opt(L, 1, "name", LUA_TSTRING, true);		\
+	const char *name = lua_tostring(L, -1);				\
+									\
+	luaxtable_reg_opt(L, 1, "revision", LUA_TNUMBER, true);		\
+	uint8_t revision = (uint8_t)lua_tointeger(L, -1);		\
+									\
+	luaxtable_reg_opt(L, 1, "family", LUA_TNUMBER, true);		\
+	unsigned short family = (unsigned short)lua_tointeger(L, -1);	\
+									\
+	unsigned proto = luaxtable_reg_opt(L, 1, "proto", LUA_TNUMBER, false) ? \
+			 (unsigned short)lua_tointeger(L, -1) : 0;	\
```

Comment:
if you are going to convert anything to integer here.. you don't need to check its type.. just do `getfield` and `tointeger`.. nil or none will be converted to `0` anyway